### PR TITLE
Added Interleave Dataset

### PR DESCRIPTION
**Pull request by vbvg2008**
`InterleaveDataset` can allow users to switch between data sources on a per-step basis. Users can also specify Interleaving pattern, and condition NumpyOps on a specific interleave source. 

This can be especially useful in multi-task training setting combined with the merge gradient feature.

### CODE REVIEW

File: fastestimator/dataset/interleave_dataset.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        # Only pass the stratify argument to the dataset(s) which have the appropriate key
+        new_datasets = [
+            to_list(ds.split(*fractions, seed=seed, stratify=stratify if stratify in ds[0] else None))
+            for ds in self.datasets
+        ]
+        num_splits = len(new_datasets[0])
+        new_datasets = [[ds[i] for ds in new_datasets] for i in range(num_splits)]
+        results = [InterleaveDataset(ds, pattern=self.pattern) for ds in new_datasets]
+        if seed is not None:
+            [ds.fe_reset_ds(seed=seed) for ds in results]
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:249]: I think that codacy is flagging this because it's constructing a list of Nones, which is an  unnecessary allocation compared to writing this as a classic for loop

File: fastestimator/dataset/combined_dataset.py [hunk 2]
```diff
@@ -22,16 +22,16 @@
 
 @traceable()
 class CombinedDataset(ConcatDataset):
-    def __init__(self, datasets: List[Dataset]) -> None:
-        """Combines a list of PyTorch Datasets,
+    """Combines a list of PyTorch Datasets,
```
Comment (TortoiseHam) [inline fastestimator/dataset/combined_dataset.py:25]:
```suggestion
    """Combines a list of PyTorch Datasets.
```

File: fastestimator/dataset/interleave_dataset.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+from fastestimator.dataset.dataset import DatasetSummary, FEDataset
+from fastestimator.util.base_util import to_list, warn
+from fastestimator.util.traceability_util import traceable
+
+
+@traceable()
+class InterleaveDataset(FEDataset):
+    """A Dataset class that can allow for a step-wise interleaving of multiple datasets. This will be useful for
+    training multi-task models when we vary dataset used on a per-step basis.
+
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:29]:
```suggestion
    """A Dataset class that can allow for a step-wise interleaving of multiple datasets.
    
    This will be useful for training multi-task models when we vary dataset used on a per-step basis.

```

File: fastestimator/dataset/interleave_dataset.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    To achieve a more complicated interleaving pattern, for example, 2 steps of `ds1` followed by 3 step of `ds2`:
+    ```python
+    dataset = InterleaveDataset(datasets=[ds1, ds2], pattern=[0, 0, 1, 1, 1])
+    dataset = InterleaveDataset(datasets={"a": ds1, "b": ds2}, pattern=["a", "a", "b", "b", "b"])
+    ```
+
+    When datasets are provided as dictionary, users can use its key as `ds_id` in other Pipeline Operators to apply
+    dataset-spefic operations (such as batching or preprocessing). For example, if we need `ds1` to go through
+    `Minmax` then form a batch of 32, and `ds2` to use `Zscore` then form a batch of 64:
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:46]:
```suggestion
    When datasets are provided as a dictionary, users can use its key as `ds_id` in other Pipeline Operators to apply
    dataset-specific operations (such as batching or preprocessing). For example, if we need `ds1` to go through
    `Minmax` then form a batch of 32, and `ds2` to use `Zscore` then form a batch of 64:
```

File: fastestimator/dataset/interleave_dataset.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                                Batch(batch_size=32, ds_id="a"),
+                                Batch(batch_size=64, ds_id="b")])
+    ```
+
+    Args:
+        datasets: List of datasets or a dictionary with key being the dataset name, value being the dataset.
+        pattern: The step-wise interleaving patterns. When datasets provided is a list, it requires list of integer
+            when `datasets` is list, requires a list of names when `datasets` is dictionary.
+    """
+    def __init__(self,
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:61]: Use typing @overload to help users catch the required dataset+pattern rule

File: fastestimator/dataset/interleave_dataset.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+                assert all([p in available_idx for p in self.pattern]), "pattern index not matching with dataset, patterns are: {}, available indexes are: {}".format(self.pattern, available_idx)
+        else:
+            raise ValueError("datasets must be either a list or dictionary")
+        # fill in default interleaving pattern
+        if self.pattern is None:
+            self.pattern = [x for x in range(len(self.datasets))]
+        # batch size information will be altered by pipeline before training.
+        self.batch_sizes = [1 for _ in self.datasets]
+        assert len(self.datasets) > 1, "require more than one dataset as input of InterleaveDataset"
+        assert len(set(self.pattern)) == len(self.datasets), "not all dataset is available in the pattern"
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:108]:
```suggestion
        assert len(set(self.pattern)) == len(self.datasets), "not all datasets are mentioned in the pattern"
```

File: fastestimator/pipeline.py [hunk 7]
```diff
@@ -223,13 +224,13 @@ def _verify_dataset(self, dataset: DataSource, **kwargs) -> bool:
         else:
             raise ValueError("Unsupported dataset type: {}".format(type(dataset)))
 
-    def _get_op_split(self, mode: str, epoch: int, ds_id: str) -> Tuple[List[NumpyOp], Batch, List[NumpyOp]]:
+    def _get_op_split(self, mode: str, epoch: int, ds_id: Iterable[str]) -> Tuple[List[NumpyOp], Batch, List[NumpyOp]]:
```
Comment (TortoiseHam) [inline fastestimator/pipeline.py:227]:
```suggestion
    def _get_op_split(self, mode: str, epoch: int, ds_id: Union[str, Iterable[str]]) -> Tuple[List[NumpyOp], Batch, List[NumpyOp]]:
```

File: fastestimator/pipeline.py [hunk 8]
```diff
@@ -591,21 +592,48 @@ def __call__(self,
...
+                ctx_ops, ctx_batch_info, ctx_batch_ops = self._get_op_split(mode=mode, epoch=epoch, ds_id={ds_id, tag} if isinstance(tag, str) else ds_id)
+                self.ctx_ops.append(ctx_ops)
+                self.ctx_batch_info.append(ctx_batch_info)
+                self.ctx_batch_ops.append(ctx_batch_ops)
+            # Decide on the batch size (this might still be ignored later if the user is using a BatchDataset)
+            self.ctx_batch_size = [ctx_batch_info.batch_size for ctx_batch_info in self.ctx_batch_info]
+            # drop_last and collate_fn for different dataset must be the same, since it is the same dataloader.
+            same_drop_last = len(set(ctx_batch_info.drop_last for ctx_batch_info in self.ctx_batch_info)) == 1
+            same_collate = len(set(ctx_batch_info.collate_fn for ctx_batch_info in self.ctx_batch_info)) == 1
+            assert same_drop_last and same_collate, "when using InterleaveDataset, the drop_last and collate behavior must be the same"
```
Comment (TortoiseHam) [inline fastestimator/pipeline.py:609]:
```suggestion
            assert same_drop_last and same_collate, "when using InterleaveDataset, the drop_last and collate behavior for all datasets must be the same"
```

File: fastestimator/pipeline.py [hunk 9]
```diff
@@ -591,21 +592,48 @@ def __call__(self,
...
+            # Interleave dataset at current scope do not support batch level, need to make sure batchops are the same
+            assert all(self.ctx_batch_ops[0] == batch_ops for batch_ops in self.ctx_batch_ops), "Current InterleaveDataset do not support different dataset behaviors after the BatchOp."
```
Comment (TortoiseHam) [inline fastestimator/pipeline.py:611]:
```suggestion
            assert all(self.ctx_batch_ops[0] == batch_ops for batch_ops in self.ctx_batch_ops), "Current InterleaveDataset does not support different dataset behaviors after the BatchOp."
```

File: fastestimator/pipeline.py [hunk 10]
```diff
@@ -591,21 +592,48 @@ def __call__(self,
...
+                self.ctx_ops.append(ctx_ops)
+                self.ctx_batch_info.append(ctx_batch_info)
+                self.ctx_batch_ops.append(ctx_batch_ops)
+            # Decide on the batch size (this might still be ignored later if the user is using a BatchDataset)
+            self.ctx_batch_size = [ctx_batch_info.batch_size for ctx_batch_info in self.ctx_batch_info]
+            # drop_last and collate_fn for different dataset must be the same, since it is the same dataloader.
+            same_drop_last = len(set(ctx_batch_info.drop_last for ctx_batch_info in self.ctx_batch_info)) == 1
+            same_collate = len(set(ctx_batch_info.collate_fn for ctx_batch_info in self.ctx_batch_info)) == 1
+            assert same_drop_last and same_collate, "when using InterleaveDataset, the drop_last and collate behavior must be the same"
+            # Interleave dataset at current scope do not support batch level, need to make sure batchops are the same
```
Comment (TortoiseHam) [inline fastestimator/pipeline.py:610]:
```suggestion
            # Interleave dataset at current scope does not support batch level, need to make sure batchops are the same
```

File: test/PR_test/integration_test/dataset/test_interleave_dataset.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+from fastestimator.dataset.numpy_dataset import NumpyDataset
+from fastestimator.op.numpyop import Batch, NumpyOp
+
+
+class Plus(NumpyOp):
+    def forward(self, data, state):
+        return data + 0.5
+
+
+class TestInterleaveDataset(unittest.TestCase):
```
Comment (TortoiseHam) [inline test/PR_test/integration_test/dataset/test_interleave_dataset.py:28]: should probably have a test to cover restore wizard, like in TestBatchDataset

File: test/PR_test/integration_test/dataset/test_interleave_dataset.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    @classmethod
+    def setUpClass(self):
+        self.data1 = {"x": [x for x in range(1, 11)], "ds_id": [0 for _ in range(10)]}
+        self.data2 = {"x": [x * 10 for x in range(1, 21)], "ds_id": [1 for _ in range(20)]}
+
+    def test_interleave_in_pipeline_default_pattern(self):
```
Comment (TortoiseHam) [inline test/PR_test/integration_test/dataset/test_interleave_dataset.py:34]: What about a test with filteredData? does it re-draw correctly from the same ds or does it mix batches?

File: fastestimator/dataset/batch_dataset.py [hunk 13]
```diff
@@ -307,11 +307,13 @@ def fe_reset_ds(self, shuffle: bool = True, *, seed: Optional[int] = None) -> No
         if self.probability:
             num_samples = num_samples * len(self.datasets)
         self.index_maps = []
-        for dataset, num_sample in zip(self.datasets, num_samples):
+        for idx, (dataset, num_sample) in enumerate(zip(self.datasets, num_samples)):
```
Comment (TortoiseHam) [inline fastestimator/dataset/batch_dataset.py:310]: we need to prevent BatchDataset from taking an InterleaveDataset as input. Same for ExtendDataset and any other meta datasets

File: fastestimator/dataset/interleave_dataset.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        Returns:
+            How many batches of data can this dataset serve per epoch. It is sum of all dataset's number of batch.
+        """
+        # first calcualte the minimum number of cycles each dataset can afford according to the repeat pattern
+        num_cycles = min(len(ds) // (f * bs) for ds, f, bs in zip(self.datasets, self.frequency, self.batch_sizes))
+        assert num_cycles > 0, "some dataset does not have enough samples for a single repeat pattern, please consider using `ExtendDataset` to increase its length"
+        # returning the sum of number of batches of each dataset
+        return sum(num_cycles * f for f in self.frequency)
+
+    def __getitem__(self, batch_idx: int) -> Dict[str, Any]:
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:161]:
```suggestion
    def __getitem__(self, batch_idx: int) -> List[Dict[str, Any]]:
```

File: fastestimator/dataset/interleave_dataset.py [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+            if isinstance(item, list):
+                batch.extend(item)
+            else:
+                batch.append(item)
+        return batch
+
+    def split(self,
+              *fractions: Union[float, int, Iterable[int]],
+              seed: Optional[int] = None,
+              stratify: Optional[str] = None) -> Union['InterleaveDataset', List['InterleaveDataset']]:
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:197]: from typing_extensions import Self

```suggestion
              stratify: Optional[str] = None) -> Union['Self', List['Self']]:
```

File: fastestimator/dataset/interleave_dataset.py [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+            datasets, tags = [], []
+            for k, v in self.datasets.items():
+                assert isinstance(k, str), "each dataset key must be a string, found: {}".format(k)
+                tags.append(k)
+                datasets.append(v)
+            self.datasets = datasets
+            self.tags = tags
+            # standardize its pattern too if it exists
+            if self.pattern:
+                assert all([p in tags for p in self.pattern]), "pattern name(s) missing in dataset dictionary, patterns are: {}, available names are: {}".format(self.pattern, tags)
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:93]: over max line length

File: fastestimator/dataset/interleave_dataset.py [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+                self.pattern = [self.tags.index(p) for p in self.pattern]
+        elif isinstance(self.datasets, list):
+            self.tags = [None for _ in self.datasets]
+            if self.pattern:
+                available_idx = [x for x in range(len(self.datasets))]
+                assert all([p in available_idx for p in self.pattern]), "pattern index not matching with dataset, patterns are: {}, available indexes are: {}".format(self.pattern, available_idx)
```
Comment (TortoiseHam) [inline fastestimator/dataset/interleave_dataset.py:99]: over max line length

File: fastestimator/dataset/interleave_dataset.py (hunk 1 above)
Comment (vbvg2008) [reply fastestimator/dataset/interleave_dataset.py:250]: alright now it's a classic for loop

File: fastestimator/dataset/interleave_dataset.py (hunk 5 above)
Comment (vbvg2008) [reply fastestimator/dataset/interleave_dataset.py:62]: done

File: test/PR_test/integration_test/dataset/test_interleave_dataset.py (hunk 11 above)
Comment (vbvg2008) [reply test/PR_test/integration_test/dataset/test_interleave_dataset.py:28]: added 👍

File: fastestimator/dataset/batch_dataset.py (hunk 13 above)
Comment (vbvg2008) [reply fastestimator/dataset/batch_dataset.py:310]: BatchDataset I agree, ExtendDataset should be ok

File: fastestimator/dataset/interleave_dataset.py (hunk 16 above)
Comment (vbvg2008) [reply fastestimator/dataset/interleave_dataset.py:94]: done

File: fastestimator/dataset/interleave_dataset.py (hunk 17 above)
Comment (vbvg2008) [reply fastestimator/dataset/interleave_dataset.py:100]: done

File: fastestimator/dataset/interleave_dataset.py [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+
+    def __len__(self) -> int:
+        """Compute the length of this dataset.
+
+        Returns:
+            How many batches of data can this dataset serve per epoch. It is sum of all dataset's number of batch.
+        """
+        # first calcualte the minimum number of cycles each dataset can afford according to the repeat pattern
+        num_cycles = min(len(ds) // (f * bs) for ds, f, bs in zip(self.datasets, self.frequency, self.batch_sizes))
+        assert num_cycles > 0, "some dataset does not have enough samples for a single repeat pattern, please consider using `ExtendDataset` to increase its length"
```
[the remark this answers is not in the document]
Comment (vbvg2008) [reply fastestimator/dataset/interleave_dataset.py:158]: done

File: fastestimator/dataset/interleave_dataset.py [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+        self.batch_sizes = None
+        self.fe_batch = None  # fe_batch member variable is used by FEDataLoader
+        self.op_datasets = None
+        self.all_fe_datasets = False
+        self.index_maps = []
+        self._standardize_args()
+        self.child_reset_fns = [ds.fe_reset_ds for ds in self.datasets if hasattr(ds, 'fe_reset_ds')]
+        self.fe_reset_ds(seed=0)  # using a seed here to ensure the same initial condition for possible seeded split.
+
+    def _standardize_args(self) -> None:
```
[the remark this answers is not in the document]
Comment (vbvg2008) [reply fastestimator/dataset/interleave_dataset.py:77]: the function is huge though, do not want to make init too long.

File: fastestimator/pipeline.py [hunk 20]
```diff
@@ -591,21 +592,48 @@ def __call__(self,
...
+            self.ctx_ops, self.ctx_batch_info, self.ctx_batch_ops = [], [], []
+            for tag in dataset.tags:
+                ctx_ops, ctx_batch_info, ctx_batch_ops = self._get_op_split(mode=mode, epoch=epoch, ds_id={ds_id, tag} if isinstance(tag, str) else ds_id)
+                self.ctx_ops.append(ctx_ops)
+                self.ctx_batch_info.append(ctx_batch_info)
+                self.ctx_batch_ops.append(ctx_batch_ops)
+            # Decide on the batch size (this might still be ignored later if the user is using a BatchDataset)
+            self.ctx_batch_size = [ctx_batch_info.batch_size for ctx_batch_info in self.ctx_batch_info]
+            # drop_last and collate_fn for different dataset must be the same, since it is the same dataloader.
+            same_drop_last = len(set(ctx_batch_info.drop_last for ctx_batch_info in self.ctx_batch_info)) == 1
```
[the remark this answers is not in the document]
Comment (vbvg2008) [reply fastestimator/pipeline.py:607]: it would have to be a re-design, I wouldn't think it's worth it for now.
